interop-reactiveStreams/shared/src/main/scala/scalaz/zio/interop/reactiveStreams/QueueSubscriber.scala

```
@@ -1,56 +1,77 @@
+    for {
+      runtime <- UIO.runtime[Any]
+      q       <- Queue.bounded[A](capacity)
+      qs      = new QueueSubscriber[A](runtime, q)
+    } yield (qs, qs)
+}
+
+private class QueueSubscriber[A](runtime: Runtime[_], q: Queue[A])
+    extends Subscriber[A]
+    with ZStream[Any, Throwable, A] {
+
+  private val capacity: Long = q.capacity.toLong
+
+  private var completed: Boolean                 = false
+  private var failed: Option[Throwable]          = None
+  private var subscription: Option[Subscription] = None
+
+  override def fold[R1 <: Any, E1 >: Throwable, A1 >: A, S]: Fold[R1, E1, A1, S] =
+    IO.succeedLazy { (s, cont, f) =>
+      def loop(s: S, demand: Long): ZIO[R1, E1, S] =
+        if (!cont(s)) subscription.fold(UIO.unit)(s => UIO(s.cancel())) *> q.shutdown.const(s)
+        else
+          q.size.flatMap { n =>
+            val empty = n <= 0
+            if (empty && completed) q.shutdown.const(s)
+            else if (empty && failed.isDefined) q.shutdown *> Task.fail(failed.get)
+            else if (empty && (demand < q.capacity))
+              subscription.fold(UIO.unit)(s => UIO(s.request(capacity - demand))) *> loop(s, capacity)
+            else q.take.flatMap(f(s, _)).flatMap(loop(_, demand - 1))
+          } <> failed.fold[Task[S]](UIO.succeed(s))(Task.fail)
+      loop(s, capacity)
     }
 
-  override def onSubscribe(subscription: Subscription): Unit = {
-    if (subscription == null) throw new NullPointerException("s was null in onSubscribe")
-    if (subscriptionOpt.isDefined) {
-      subscription.cancel()
-    } else {
-      subscriptionOpt = Some(subscription)
-      runtime.unsafeRunAsync_(q.awaitShutdown *> UIO(subscription.cancel()))
-      subscription.request(q.capacity.toLong)
+  override def onSubscribe(s: Subscription): Unit = {
+    if (s == null) throw new NullPointerException("s was null in onSubscribe")
+    synchronized {
```

Feels weird to use synchronized here. It is only a guard against misuse of the Subscriber. `onSubscribe` should only be called once in the lifetime of a `QueueSubscriber`. Synchronizing here saves us a `Promise.await` in each call to `loop` in `fold`.

---

I'd just throw an exception here if `onSubscribe` is somehow called twice and let the fiber die. The graceful handling of that here might mask implementation errors.

---

According to the spec, we must cancel the subscription and must not fail, so that is not possible.

---

I see. Ok then. I'm good with the use of `synchronized` here.